File: src/checks/webhooks.ts
import { createHmac } from 'crypto';
import { Check, CheckContext, CheckResult } from '../types.js';
import { Drip } from '../drip-client.js';

// Local helper to generate webhook signature (avoids ESM/CJS issues with SDK)
function generateTestSignature(payload: string, secret: string): string {
  const timestamp = Math.floor(Date.now() / 1000);
  const signaturePayload = `${timestamp}.${payload}`;
  const signature = createHmac('sha256', secret)
    .update(signaturePayload)
    .digest('hex');
  return `t=${timestamp},v1=${signature}`;
}

export const webhookSignCheck: Check = {
  name: 'Webhook Sign',
  description: 'Test webhook signature generation and verification',
  quick: true,
  async run(_ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();

    try {
      const testPayload = JSON.stringify({
        type: 'charge.succeeded',
        data: { chargeId: 'chg_test123' },
      });
      const testSecret = 'whsec_test_secret';

      // Generate a valid signature locally
      const validSignature = generateTestSignature(testPayload, testSecret);

      // Verify the signature using the SDK's async method
      const isValid = await Drip.verifyWebhookSignatureAsync(
        testPayload,
        validSignature,
        testSecret
      );

      const duration = performance.now() - start;

      if (isValid) {
        return {
          name: 'Webhook Sign',
          success: true,
          duration,
          message: 'Signature valid',
        };
      } else {
        return {
          name: 'Webhook Sign',
          success: false,
          duration,
          message: 'Valid signature was rejected',
          suggestion: 'Webhook signature generation/verification is broken',
        };
      }
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as Error;
      return {
        name: 'Webhook Sign',
        success: false,
        duration,
        message: err.message || 'Signature verification failed',
      };
    }
  },
};

export const webhookVerifyCheck: Check = {
  name: 'Webhook Verify',
  description: 'Test invalid signature rejection',
  quick: true,
  async run(_ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();

    try {
      const testPayload = JSON.stringify({
        type: 'charge.succeeded',
        data: { chargeId: 'chg_test123' },
      });
      const testSecret = 'whsec_test_secret';
      const invalidSignature = 'invalid_signature';

      const isValid = await Drip.verifyWebhookSignatureAsync(
        testPayload,
        invalidSignature,
        testSecret
      );

      const duration = performance.now() - start;

      if (!isValid) {
        return {
          name: 'Webhook Verify',
          success: true,
          duration,
          message: 'Invalid sig rejected',
        };
      } else {
        return {
          name: 'Webhook Verify',
          success: false,
          duration,
          message: 'Invalid signature was accepted',
          suggestion: 'Webhook signature verification is not working correctly',
        };
      }
    } catch (error) {
      const duration = performance.now() - start;
      // Rejection via exception is also valid
      return {
        name: 'Webhook Verify',
        success: true,
        duration,
        message: 'Invalid sig rejected',
      };
    }
  },
};

File: python/checks/webhooks.py
"""Webhook checks."""
import uuid
import hmac
import hashlib
from ..types import Check, CheckContext, CheckResult
from ..drip_client import create_client


async def _webhook_sign_check(ctx: CheckContext) -> CheckResult:
    """Create a webhook and get signing secret."""
    try:
        client = create_client(ctx.api_key, ctx.api_url)

        # Check if SDK supports webhooks
        if not hasattr(client, 'create_webhook'):
            return CheckResult(
                name="webhook_sign",
                success=True,
                duration=0,
                message="Webhook creation not available in SDK",
                details="Skipping webhook tests"
            )

        webhook = client.create_webhook(
            url=f"https://example.com/webhook/{uuid.uuid4().hex}",
            events=["charge.succeeded", "customer.balance.low"],
            description="Health check webhook"
        )

        # Store for verify check
        ctx.webhook_id = webhook.id
        ctx.webhook_secret = getattr(webhook, 'secret', None)

        return CheckResult(
            name="webhook_sign",
            success=True,
            duration=0,
            message=f"Created webhook {webhook.id}",
            details="Secret obtained for verification"
        )
    except Exception as e:
        return CheckResult(
            name="webhook_sign",
            success=False,
            duration=0,
            message=f"Failed to create webhook: {e}"
        )


webhook_sign_check = Check(
    name="webhook_sign",
    description="Create webhook and get secret",
    run=_webhook_sign_check,
    quick=True
)


async def _webhook_verify_check(ctx: CheckContext) -> CheckResult:
    """Verify webhook signature validation."""
    if not ctx.webhook_secret:
        return CheckResult(
            name="webhook_verify",
            success=True,
            duration=0,
            message="No webhook secret available",
            details="Webhook sign check may have been skipped"
        )

    try:
        # Create test payload and signature
        test_payload = b'{"event": "test", "data": {}}'

        # Generate valid signature
        signature = hmac.new(
            ctx.webhook_secret.encode(),
            test_payload,
            hashlib.sha256
        ).hexdigest()

        # Try to use SDK's verify function if available
        try:
            from drip import verify_webhook_signature
            is_valid = verify_webhook_signature(
                payload=test_payload,
                signature=signature,
                secret=ctx.webhook_secret
            )
        except ImportError:
            # Fallback: manually verify
            expected_sig = hmac.new(
                ctx.webhook_secret.encode(),
                test_payload,
                hashlib.sha256
            ).hexdigest()
            is_valid = hmac.compare_digest(signature, expected_sig)

        if is_valid:
            return CheckResult(
                name="webhook_verify",
                success=True,
                duration=0,
                message="Signature verification working"
            )
        else:
            return CheckResult(
                name="webhook_verify",
                success=False,
                duration=0,
                message="Signature verification failed unexpectedly"
            )
    except Exception as e:
        return CheckResult(
            name="webhook_verify",
            success=False,
            duration=0,
            message=f"Webhook verify failed: {e}"
        )
    finally:
        # Cleanup webhook
        if ctx.webhook_id:
            try:
                client = create_client(ctx.api_key, ctx.api_url)
                if hasattr(client, 'delete_webhook'):
                    client.delete_webhook(ctx.webhook_id)
            except Exception:
                pass


webhook_verify_check = Check(
    name="webhook_verify",
    description="Verify webhook signature",
    run=_webhook_verify_check,
    quick=True
)
